### ISSUE TITLE
Make benchmark suite faster

### ISSUE THREAD

**Issue by pv** [scipy/scipy#9225]
The full benchmark suite currently takes 7 hours to run. I suspect it doesn't need to be this slow.

There are a couple of things to improve:

- [ ] Do all the benchmarks make sense? Is there a well-defined need to e.g. benchmark over all possible parameter combinations?
- [ ] Inspect slow benchmarks: https://pv.github.io/scipy-bench/#summarylist?sort=1&dir=desc --- if it takes very long, does it make sense to benchmark, or are the parameters chosen not appropriate?
- [ ] Inspect slow setup routines shared by multiple benchmarks. Here's a script for timing setup routines: https://gist.github.com/pv/9f51381ec8a488c34920749d0f00c43e

Similarly as with tests, slow benchmarks could be marked with a `slow` attribute.
Asv doesn't have a way to do this, but we could use an environment variable
and raise NotImplementedError() from `setup` routines for slow ones.

**Comment by perimosocordiae** [comment]
The top slowest benchmark on that list is `spatial.Xdist.time_pdist(1000, 'dice')` at almost two minutes / run, but I can't reproduce that locally:

```
In [2]: import scipy.spatial.distance as dist

In [4]: np.random.seed(123)

In [5]: points = np.random.random_sample((1000, 3))

In [11]: %timeit dist.pdist(points, 'euclidean', w=None)
1000 loops, best of 3: 1.51 ms per loop

In [12]: %timeit dist.pdist(points, 'yule', w=None)
100 loops, best of 3: 3.31 ms per loop

In [13]: %timeit dist.pdist(points, 'dice', w=None)
100 loops, best of 3: 3.33 ms per loop
```

**Comment by pv** [comment]
For me (on intel i7 laptop --- *not* particularly slow machine):
```
In [1]: import numpy as np
In [2]: from scipy.spatial import distance
In [3]: points = np.random.random_sample((1000, 3))
In [4]: %timeit -n1 -r1 distance.pdist(points, 'dice', w=None)
9.23 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
In [5]: %timeit -n1 -r1 dist.pdist(points, 'yule', w=None)
11.8 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
In [6]: import scipy
In [7]: scipy.__version__
'1.2.0.dev0+5f307ce'
```
That's then is probably multiplied by ~10x on the benchmark machine with slow Atom processor.

But something has to be wrong if you get 3ms?

**Comment by pv** [comment]
There's a big performance regression somewhere after Scipy 0.19.1 --- it's fast for me when on 0.19.1.

**Comment by perimosocordiae** [comment]
Yep, my results were from a machine with an older scipy version. 

With your recent changes, how long does the benchmark suite take now?

**Comment by pv** [comment]
On the benchmark machine it's 3.5 hours per commit, as of cc75a833 (instead of 7h as in ee2f3951).

**Comment by larsoner** [comment]
@pv can you update https://pv.github.io/scipy-bench/#summarylist?sort=1&dir=desc ? The first bunch of entries are still `spatial.Xdist.time_pdist`. If it is up to date, we should probably drop the `1000` condition because that's still taking a lot of time.

**Comment by pv** [comment]
It updates daily, and probably considers a single measurement point a fluke.
